java/ql/src/Security/CWE/CWE-346/UnvalidatedCors.ql

```diff
@@ -49,6 +57,21 @@ class CorsOriginConfig extends TaintTracking::Configuration {
       sink.asExpr() = corsheader.getArgument(1)
     )
   }
+
+  /*
+   * This should ideally check, the origin being validated against a list/array-list.
+   * or function being used to validate the origin, which has a flow from its parameter to any of the CorsProbableCheckAccess functions
+   */
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    node.asExpr() = any(CorsProbableCheckAccess ma).getAnArgument()
+    or
+    exists(MethodAccess ma, CorsProbableCheckAccess ca |
+      ma.getMethod().calls(ca.getMethod()) and
+      DataFlow::localExprFlow(ma.getMethod().getAParameter().getAnAccess(), ca.getAnArgument()) and
+      (node.asExpr() = ma.getAnArgument() or node.asExpr() = ma.getAnArgument().getAChildExpr())
+    )
+  }
```

Comment:
```suggestion
  override predicate isSanitizer(DataFlow::Node node) {
    exists(CorsProbableCheckAccess check |
      DataFlow::localFlow(node, DataFlow::exprNode(check.getAnArgument()))
    ) or
    exists(MethodAccess wrapperAccess, Method wrapper, int i, CorsProbableCheckAccess check |
      node.asExpr() = wrapperAccess.getArgument(i) and
      wrapperAccess.getMethod() = wrapper and
      DataFlow::localFlow(DataFlow::parameterNode(wrapper.getParameter(i)), DataFlow::exprNode(check.getAnArgument()))
    )
  }
```

Reply:
@pwntester is there a difference between doing  `localFlow` as to invoking `localExprFlow` with expressions.
Quick question on a couple of other things.
* `wrapperAccess.getMethod() = wrapper` would get all methods as opposed to `wrapperAccess.getMethod().calls(check.getMethod())`, aren't we interested only in reducing the search space to wrapper methods calling `CorsProbableCheckAccess` methods ?
* Correct me if i'm wrong here, with `check.getAnArgument()` in `DataFlow::localFlow(DataFlow::parameterNode(wrapper.getParameter(i)), DataFlow::exprNode(check.getAnArgument()))` , this doesn't catch cases like the one below (from uaarkoti/cors-filter), as `getAnArgument` returns `origin.trim()` as the argument expression, hence i had included the `getAChildExpr` which gets the `origin` in this case and able to check/complete for a local dataflow check.
```
if (origin != null && isAllowed(origin.trim())) 
```
*  Probably a dumb question,  on the `int i` stuff in `exists`, how does `exists` know, what is the range of `i` it should try upto in `getArgument(i)` ?

Reply:
> aren't we interested only in reducing the search space to wrapper methods calling CorsProbableCheckAccess methods ?

That should be the case since we are looking for wrapper methods with a parameter which flows into a `CorsProbableCheckAccess`:

```
DataFlow::localFlow(DataFlow::parameterNode(wrapper.getParameter(i)), DataFlow::exprNode(check.getAnArgument()))
```

> this doesn't catch cases like the one below (from `uaarkoti/cors-filter`), as `getAnArgument` returns `origin.trim()` as the argument expression

`check.getAnArgument()` is matching the call to `Collection.contains()` and `isAllowed(origin.trim())` is the wrapper call. Standard TaintTracking should track the taint till the `origin.trim()` return value. Which we are later marking as the source of our wrapper intra-procedural dataflow so it should be ok

> also why not getAnArgument?

With `getAnArgument` there may be a mismatch between the tainted expression reaching the wrapper and the source of the second dataflow. For example, having `isAllowed(tainted, value)` and then an `isAllowed` method like

```
public boolean isAllowed(foo, value) {
  // use foo for something else
  return allowedOrigins.contains(value);
}
```

Reply:
@pwntester thanks for clarification.

> check.getAnArgument() is matching the call to Collection.contains() and isAllowed(origin.trim()) is the wrapper call. Standard TaintTracking should track the taint till the origin.trim() return value. Which we are later marking as the source of our wrapper intra-procedural dataflow so it should be ok.

I had the same initial assumption on taint tracking till return value, until i was evaluating the sanitizer against (uaarkoti/cors-filter), this change returns a result for me which is a FP like said, can you cross check this as well if possible ?. If this is fine we can merge the suggestions then.

Reply:
I see what you mean now, the problem is not `check.getAnArgument()` but `node.asExpr() = wrapperAccess.getArgument(i)` since that will mark `origin.trim()` as the non-propagating node instead of `origin`. This version removes many other FPs so let me take a look and see how we can get the best of the two versions

Reply:
Updated the suggestion above to use two local taint tracking flows. This works several cases were we got FPs (including `uaarkoti/cors-filter`):
https://lgtm.com/query/9089103057514981051/
https://lgtm.com/query/6833608705212745163/
https://lgtm.com/query/2743352730880473023/

Reply:
@pwntester awesome, yup my initial idea was to get the `node.asExpr() = ma.getAnArgument().getAChildExpr()` which ideally would take care of this scenario by getting `origin` as child expr, but taint tracking that (condition) node is a way better idea :)